tests/Service/ContributionsService/FindByContributionToProgrammeTest.php

```diff
@@ -2,49 +2,55 @@
+
 class FindByContributionToProgrammeTest extends AbstractContributionsServiceTest
 {
-    public function testFindByContributionToProgrammeDefaultPagination()
+    /**
+     * @dataProvider paginationProvider
+     */
+    public function testPagination($expectedLimit, $expectedOffset, array $paginationParams)
     {
-        $dbId = 1;
-        $programme = $this->mockEntity('Programme', $dbId);
-        $dbData = [['pid' => 'b00swyx1'], ['pid' => 'b010t150']];
+        $programme = $this->createConfiguredMock(Programme::class, ['getDbId' => 1]);
 
         $this->mockRepository->expects($this->once())
             ->method('findByContributionTo')
-            ->with([$dbId], 'programme', false, 300, 0)
-            ->willReturn($dbData);
+            ->with([$programme->getDbId()], 'programme', false, $expectedLimit, $expectedOffset);
 
-        $result = $this->service()->findByContributionToProgramme($programme);
-        $this->assertEquals($this->contributionsFromDbData($dbData), $result);
+        $this->service()->findByContributionToProgramme($programme, ...$paginationParams);
     }
 
-    public function testFindByContributionToProgrammeCustomPagination()
+    public function paginationProvider(): array
     {
-        $dbId = 1;
-        $programme = $this->mockEntity('Programme', $dbId);
-        $dbData = [['pid' => 'b00swyx1'], ['pid' => 'b010t150']];
-
-        $this->mockRepository->expects($this->once())
-            ->method('findByContributionTo')
-            ->with([$dbId], 'programme', false, 5, 10)
-            ->willReturn($dbData);
-
-        $result = $this->service()->findByContributionToProgramme($programme, 5, 3);
-        $this->assertEquals($this->contributionsFromDbData($dbData), $result);
+        return [
+          'CASE: default' => [300, 0, []],
+          'CASE: custom' => [3, 12, [3, 5]],
+        ];
     }
 
-    public function testFindByContributionToProgrammeWithNonExistantDbId()
+    /**
+     * @dataProvider resultsDbProvider
+     */
+    public function testReturnResultsFound(array $expectedPids, array $fakeDbContributions)
     {
-        $dbId = 999;
-        $programme = $this->mockEntity('Programme', $dbId);
+        $this->mockRepository->method('findByContributionTo')->willReturn($fakeDbContributions);
 
-        $this->mockRepository->expects($this->once())
-            ->method('findByContributionTo')
-            ->with([$dbId], 'programme', false, 5, 10)
-            ->willReturn([]);
+        $contributions = $this->service()->findByContributionToProgramme($this->createMock(Programme::class));
 
-        $result = $this->service()->findByContributionToProgramme($programme, 5, 3);
-        $this->assertEquals([], $result);
+        $this->assertCount(count($fakeDbContributions), $contributions);
+        $this->assertContainsOnly(Contribution::class, $contributions);
+        $this->assertEquals($expectedPids, $this->extractPids($contributions));
```

Comment:
I'm not overly keen on asserting the return value of a method in this test class equals a whole array. Is there a way to do this that requires less data manipulation?

Reply:
I aggree, but not sure if is possible.....we can add multiple assets in rack for each item, can be more transparent, but it also creates more noise with so many asserts....

My idea is to keep it as such. Basically with this we are testing that after creating the entity models (produced really by the mapper but in the tests for us) none of our results are discarded, reordered, etc, so we can continue focusing on the output.
@BPScott any thought?

Reply:
Maybe a foreach?

Reply:
Actually that is extractPids:
```php
    protected function extractPids(array $contributions): array
    {
        return array_map(
            function ($contribution) {
                return (string) $contribution->getPid();
            },
            $contributions
        );
    }
```

Reply:
But that hides what is happening.

Reply:
maybe we can create something more generic:

```php
function extractProperty(string $property, array $items){
   return ......
}
```

At least this avoid to have extractPids, extractDbIds, ......

Reply:
I don't like extracting anything as part of tests